Core/MIPS/ARM64/Arm64CompLoadStore.cpp

```diff
@@ -0,0 +1,385 @@
+#define _RT MIPS_GET_RT(op)
+#define _RD MIPS_GET_RD(op)
+#define _FS MIPS_GET_FS(op)
+#define _FT MIPS_GET_FT(op)
+#define _FD MIPS_GET_FD(op)
+#define _SA MIPS_GET_SA(op)
+#define _POS  ((op>> 6) & 0x1F)
+#define _SIZE ((op>>11) & 0x1F)
+#define _IMM16 (signed short)(op & 0xFFFF)
+#define _IMM26 (op & 0x03FFFFFF)
+
+// All functions should have CONDITIONAL_DISABLE, so we can narrow things down to a file quickly.
+// Currently known non working ones should have DISABLE.
+
+// #define CONDITIONAL_DISABLE { Comp_Generic(op); return; }
+#define CONDITIONAL_DISABLE ;
+#define DISABLE { Comp_Generic(op); return; }
+
+namespace MIPSComp
+{
+	using namespace Arm64Gen;
+	using namespace Arm64JitConstants;
+	
+	// Destroys SCRATCH2
+	void Arm64Jit::SetScratch1ToEffectiveAddress(MIPSGPReg rs, s16 offset) {
+		if (offset) {
+			ADDI2R(SCRATCH1, gpr.R(rs), offset, SCRATCH2);
+		} else {
+			MOV(SCRATCH1, gpr.R(rs));
+		}
+	}
+
+	void Arm64Jit::SetCCAndSCRATCH1ForSafeAddress(MIPSGPReg rs, s16 offset, ARM64Reg tempReg, bool reverse) {
+		SetScratch1ToEffectiveAddress(rs, offset);
+		// TODO
+	}
+
+	void Arm64Jit::Comp_ITypeMemLR(MIPSOpcode op, bool load) {
+		CONDITIONAL_DISABLE;
+		int offset = (signed short)(op & 0xFFFF);
+		MIPSGPReg rt = _RT;
+		MIPSGPReg rs = _RS;
+		int o = op >> 26;
+
+		DISABLE;
+		// TODO: For some reason I can't get this to work on ARM64.
+		if (!js.inDelaySlot && false) {
+			// Optimisation: Combine to single unaligned load/store
+			bool isLeft = (o == 34 || o == 42);
+			MIPSOpcode nextOp = Memory::Read_Instruction(js.compilerPC + 4);
+			// Find a matching shift in opposite direction with opposite offset.
+			if (nextOp == (isLeft ? (op.encoding + (4 << 26) - 3)
+				: (op.encoding - (4 << 26) + 3))) {
+				EatInstruction(nextOp);
+				nextOp = MIPSOpcode(((load ? 35 : 43) << 26) | ((isLeft ? nextOp : op) & 0x03FFFFFF)); //lw, sw
+				Comp_ITypeMem(nextOp);
+				return;
+			}
+		}
+
+		u32 iaddr = gpr.IsImm(rs) ? offset + gpr.GetImm(rs) : 0xFFFFFFFF;
+		bool doCheck = false;
+		FixupBranch skip;
+
+		if (gpr.IsImm(rs) && Memory::IsValidAddress(iaddr)) {
+			u32 addr = iaddr & 0x3FFFFFFF;
+			// Need to initialize since this only loads part of the register.
+			// But rs no longer matters (even if rs == rt) since we have the address.
+			gpr.MapReg(rt, load ? MAP_DIRTY : 0);
+			gpr.SetRegImm(SCRATCH1, addr & ~3);
+
+			u8 shift = (addr & 3) * 8;
+
+			switch (o) {
+			case 34: // lwl
+				LDR(SCRATCH1, MEMBASEREG, SCRATCH1);
+				ANDI2R(gpr.R(rt), gpr.R(rt), 0x00ffffff >> shift, SCRATCH2);
+				ORR(gpr.R(rt), gpr.R(rt), SCRATCH1, ArithOption(gpr.R(rt), ST_LSL, 24 - shift));
+				break;
+
+			case 38: // lwr
+				LDR(SCRATCH1, MEMBASEREG, SCRATCH1);
+				ANDI2R(gpr.R(rt), gpr.R(rt), 0xffffff00 << (24 - shift), SCRATCH2);
+				ORR(gpr.R(rt), gpr.R(rt), SCRATCH1, ArithOption(gpr.R(rt), ST_LSR, shift));
+				break;
+
+			case 42: // swl
+				LDR(SCRATCH2, MEMBASEREG, SCRATCH1);
+				// Don't worry, can't use temporary.
```

Comment:
Is this still true (and below) for arm64?  Haven't thought about it much, but I know imms are represented differently.

-[Unknown]